src/engine/renderer/tr_image_ktx.cpp

```diff
@@ -38,143 +42,206 @@ struct KTX_header_t {
-	    hdr->pixelDepth != 0 ) {
-		ri.FS_FreeFile( hdr );
-		return;
+uint32_t GetImageSize( const byte *ktxPosition, bool needReverseBytes ) {
+	uint32_t imageSize = *(reinterpret_cast<const uint32_t *>(ktxPosition));
+
+	if( needReverseBytes )
+		imageSize = Swap32(imageSize);
+
+	return PAD(imageSize, 4U);
+}
+
+bool IsNonArrayCubemapTexture( const KTX_header_t *hdr ) {
+	return hdr->numberOfArrayElements == 0 && hdr->numberOfFaces == 6;
+}
+
+bool IsValidKTXFileStreamPosition( const byte *position, size_t ktxSize, const byte *ktxData ) {
+	return position >= ktxData && position < ktxData + ktxSize;
+}
+
+bool LoadInMemoryKTX( const char *name, void *ktxData, size_t ktxSize,
+    			     byte **data, int *width, int *height, int *numLayers,
+    			     int *numMips, int *bits ) {
+	auto *hdr{ static_cast<KTX_header_t *>(ktxData) };
+
+	if( !IsValidKTXHeader( hdr, ktxSize ) ) {
+		Log::Warn("KTX texture '%s' is not in KTX format", name);
+		return false;
+	}
+
+	bool needReverseBytes{false};
+	if( !TryApplyKTXHeaderEndianness( hdr, needReverseBytes ) ) {
+		Log::Warn("KTX texture '%s' has unknown endianness value %u", name,
+		hdr->endianness);
+		return false;
+	}
+
+	if( !IsSupportedKTXFormat( hdr ) ) {
+		Log::Warn("KTX texture '%s' may be valid, but not supported", name);
+		return false;
+	}
+
+	if( !TryParseInternalFormatBits( hdr->glInternalFormat, *bits ) ) {
+		Log::Warn("KTX texture '%s' has unsupported glInternalFormat value %u",
+		name, hdr->glInternalFormat);
+		return false;
 	}
 
 	*width = hdr->pixelWidth;
 	*height = hdr->pixelHeight;
 	*numMips = hdr->numberOfMipmapLevels;
 	*numLayers = hdr->numberOfFaces == 6 ? 6 : 0;
 
-	ptr = (byte *)(hdr + 1);
-	ptr += hdr->bytesOfKeyValueData;
-	size = 0;
-	for(unsigned i = 0; i < hdr->numberOfMipmapLevels; i++ ) {
-		imageSize = *((uint32_t *)ptr);
-		if( hdr->endianness == KTX_endianness_reverse )
-			imageSize = Swap32( imageSize );
-		imageSize = PAD( imageSize, 4 );
-
-		size += imageSize * hdr->numberOfFaces;
-		ptr += 4 + imageSize;
+	byte *firstImageDataPtr{ (byte *)(hdr + 1) + hdr->bytesOfKeyValueData };
+	if ( !IsValidKTXFileStreamPosition( firstImageDataPtr, ktxSize, static_cast<const byte *>(ktxData) ) ) {
+		Log::Warn("KTX texture '%s' has bad bytesOfKeyValueData or texture data", name);
+		return false;
 	}
+	byte *ptr{ firstImageDataPtr };
+	size_t totalImageSize{ 0 };
 
-	ptr = (byte *)(hdr + 1);
-	ptr += hdr->bytesOfKeyValueData;
-	data[ 0 ] = (byte *)ri.Z_Malloc( size );
+	// For most textures imageSize is the number of bytes of pixel data in the
+	// current LOD level. The exception is non-array cubemap textures where
+	// imageSize is the number of bytes in each face of the texture for the
+	// current LOD level.
+	const uint32_t mipmapLevelCoefficient{
+		IsNonArrayCubemapTexture(hdr) ? hdr->numberOfFaces : 1U};
+
+	for(uint32_t i{ 0 }; i < hdr->numberOfMipmapLevels; i++) {
+		const uint32_t imageSize{ GetImageSize( ptr, needReverseBytes ) };
```

Comment:
check that the pointers to image size are inside the file bounds?

Reply:
done.